platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingArtifactVariantSelector.java

```diff
@@ -138,20 +147,105 @@ private static List<TransformedVariant> tryDisambiguate(
 
         assert !matches.isEmpty();
 
-        List<TransformedVariant> differentTransforms = new ArrayList<>(1);
+        /*
+         At this point, we may have ambiguity.  More than 1 match is okay IFF there are only 1 truly distinct
+         transform chains in the set of matches.
+
+         As chains of A -> B -> C -> D and A -> C -> B -> D are merely re-sequencings of the same chain they are
+         not truly distinct.  This is fine , Gradle will just arbitrarily pick one, as the different order
+         that steps are run is PROBABLY not meaningful - the same work will be done.
+
+         However, if matches contains chains of A -> B -> C and A -> D -> C this is NOT okay!  Even if they end up
+         producing a C with the same exact attributes, they represent DIFFERENT work being done, selected
+         arbitrarily, and could have significant impact on the overall build time.  The build author should address
+         this ambiguity.
+        */
 
-        // Choosing the last candidate here is arbitrary.
-        TransformedVariant last = matches.get(matches.size() - 1);
-        differentTransforms.add(last);
+        // These are arbitrarily selected chains that produce incompatible results
+        List<TransformedVariant> chainsProducingDistinctResults = new ArrayList<>(1);
 
-        // Find any other candidate which does not match with the last candidate.
+        // Choosing the last candidate as the winner here is arbitrary, but should be preserved for backwards compatibility
+        TransformedVariant arbitrarilySelectedWinningChain = matches.get(matches.size() - 1);
+        chainsProducingDistinctResults.add(arbitrarilySelectedWinningChain);
+
+        // Find all other candidate chains which do not result in a mutually compatible result with the arbitrary selection
         for (int i = 0; i < matches.size() - 1; i++) {
             TransformedVariant current = matches.get(i);
-            if (!matcher.areMutuallyCompatible(current.getAttributes(), last.getAttributes())) {
-                differentTransforms.add(current);
+            if (!matcher.areMutuallyCompatible(current.getAttributes(), arbitrarilySelectedWinningChain.getAttributes())) {
```

Comment:
There is a lot of new logic here in a class that is already quite complex. Imagine we can simplify this into a single iteration while reducing code complexity

Reply:
I think the major change to this code I would make is to compare whether a given candidate (`current`) in this loop was truly distinct at the time of iterating in this loop, not after the fact.

I think we could solve this with a Map<TransformedVariant, List<TransformedVaraint>> where the key is distinct "arbitrarily selected" variant, and the values are a list of each variant that produces compatible attributes, but with a non compatible chain.

I haven't spent the time atm to read all the logic in here, but my initial reaction is we might be able to simplify this a bit